capa/features/extractors/dnfile/extractor.py

```diff
@@ -127,26 +135,100 @@ def get_functions(self) -> Iterator[FunctionHandle]:
+        for fh in methods.values():
+            # calculate basic block leaders where,
+            #   1. The first instruction of the intermediate code is a leader
+            #   2. Instructions that are targets of unconditional or conditional jump/goto statements are leaders
+            #   3. Instructions that immediately follow unconditional or conditional jump/goto statements are considered leaders
+            #   https://www.geeksforgeeks.org/basic-blocks-in-compiler-design/
+
+            leaders: Set[int] = set()
+            for idx, insn in enumerate(fh.inner.instructions):
+                if idx == 0:
+                    # add #1
+                    leaders.add(insn.offset)
+
+                if any((insn.is_br(), insn.is_cond_br(), insn.is_leave())):
+                    # add #2
+                    leaders.add(insn.operand)
+                    # add #3
+                    try:
+                        leaders.add(fh.inner.instructions[idx + 1].offset)
+                    except IndexError:
+                        # may encounter branch at end of method
+                        continue
+
+            # build basic blocks using leaders
+            bb_curr: Optional[DnBasicBlock] = None
+            for idx, insn in enumerate(fh.inner.instructions):
+                if insn.offset in leaders:
+                    # new leader, new basic block
+                    bb_curr = DnBasicBlock(instructions=[insn])
+                    fh.ctx["blocks"].append(bb_curr)
+                    continue
+
+                assert bb_curr is not None
+                bb_curr.instructions.append(insn)
+
+            # create mapping of first instruction to basic block
+            bb_map: Dict[int, DnBasicBlock] = {}
+            for bb in fh.ctx["blocks"]:
+                if len(bb.instructions) == 0:
+                    # TODO: consider error?
+                    continue
+                bb_map[bb.instructions[0].offset] = bb
+
+            # connect basic blocks
+            for idx, bb in enumerate(fh.ctx["blocks"]):
+                if len(bb.instructions) == 0:
+                    # TODO: consider error?
+                    continue
+
+                last = bb.instructions[-1]
+
+                # connect branches to other basic blocks
+                if any((last.is_br(), last.is_cond_br(), last.is_leave())):
+                    bb_branch: Optional[DnBasicBlock] = bb_map.get(last.operand, None)
+                    if bb_branch is not None:
+                        # TODO: consider None error?
+                        bb.succs.append(bb_branch)
+                        bb_branch.preds.append(bb)
+
+                if any((last.is_br(), last.is_leave())):
+                    # no fallthrough
+                    continue
+
+                # connect fallthrough
+                try:
+                    bb_next: DnBasicBlock = fh.ctx["blocks"][idx + 1]
+                    bb.succs.append(bb_next)
```

Comment:
Does this always hold just based on the next index?

Reply:
Here is my thought process (please check my sanity 🙃):

- We process instructions sequentially resulting in leaders getting processed sequentially
- Each time we encounter a leader we create a new basic block and add the block to our list resulting in basic blocks getting processed sequentially
- Therefore, if a basic block has a fallthrough successor we can assume the successor is the next basic block in our list

Reply:
That makes sense and should go in there as a comment.

What about BBs that do not return (IDA calls them `fcb_noret` and `fcb_enoret`)? They occur for example on `call abort`, `RtlFailFast`, or potentially on exception handling.

Another corner case from IDA at the `call 0x100304ae`:
<img width="1059" alt="2023-03-01_09-42-28_ida" src="https://user-images.githubusercontent.com/17606537/222088147-7f53e7e8-7b69-462a-a69c-841a12a2201f.png">
```
ID: 8, Start: 0x1003045e, Last instruction: 0x10030465, Size: 12, Type: fcb_normal
  -> ID: 16, Start: 0x100304ae, Last instruction: 0x100304b7, Size: 10, Type: fcb_normal
```

So we should handle `call` and `jmp` at the end of BBs. And maybe exception handling could throw us of?

Reply:
agree the above should be in a comment, just for clarity.

i'm not sure whether or not .NET supports non-returning functions. its possibly that the runtime still requires the `ret` even if its dead code. we should try to figure this out and definitely document the results here.

Reply:
other assumptions we should probably document (and please verify):
  - .NET doesn't support tail calls, i.e. `jmp` to other routine
  - .NET doesn't support shared function chunks, i.e.  two different entry points to the same function body